### PR TITLE
Final bringup nomap

### PR DESCRIPTION
**Pull request by jr1221**
## Changes

_Explanation of changes goes here_

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)

### CODE REVIEW

File: Core/Src/can_messages.c [hunk 1]
```diff
@@ -241,35 +241,23 @@ void send_shutdown_ctrl_message(uint8_t mpe_state)
...
+	bitstream_add(&cell_voltage_msg, min_voltage.chipIndex, 4);
+	bitstream_add(&cell_voltage_msg, min_voltage.cellNum, 4);
+	bitstream_add(&cell_voltage_msg, avg_voltage * 10000, 16);
 
 	can_msg_t msg = { .id = CELL_DATA_CANID,
 			  .len = CELL_DATA_SIZE,
 			  .data = { 0 } };
 
-	memcpy(msg.data, &cell_data_msg_data, sizeof(cell_data_msg_data));
+	memcpy(msg.data, &bitstream_data, sizeof(bitstream_data));
```
Comment (caiodasilva2005) [inline Core/Src/can_messages.c:260]: do we know if this was working? because bitstream_data is already a pointer and we are passing in its address into the memcpy here, which should not work

File: Core/Src/can_messages.c [hunk 2]
```diff
@@ -303,35 +291,23 @@ void send_segment_volt_message(acc_data_t *bmsdata)
...
+	bitstream_add(&cell_voltage_msg, min_temp.chipIndex, 4);
+	bitstream_add(&cell_voltage_msg, min_temp.cellNum, 4);
+	bitstream_add(&cell_voltage_msg, avg_temp * 100, 16);
 
 	can_msg_t msg = { .id = CELL_TEMP_CANID,
 			  .len = CELL_TEMP_SIZE,
 			  .data = { 0 } };
 
-	memcpy(msg.data, &cell_temp_msg_data, sizeof(cell_temp_msg_data));
+	memcpy(msg.data, &bitstream_data, sizeof(bitstream_data));
```
Comment (caiodasilva2005) [inline Core/Src/can_messages.c:310]: same thing here

File: Core/Src/compute.c [hunk 3]
```diff
@@ -68,7 +68,7 @@ int8_t compute_measure_temp(float *temp, float *humidity)
 
 void compute_set_fault(bool fault_state)
 {
-	HAL_GPIO_WritePin(FAULT_MCU_GPIO_Port, FAULT_MCU_Pin, !fault_state);
+	HAL_GPIO_WritePin(FAULT_MCU_GPIO_Port, FAULT_MCU_Pin, fault_state);
```
Comment (caiodasilva2005) [inline Core/Src/compute.c:71]: isn't fault supposed to be active low? This could have been changed, but just wanna confirm. I know we were getting bms faults tho

**Review by caiodasilva2005 — CHANGES_REQUESTED** [review]

File: Core/Inc/bmsConfig.h [hunk 4]
```diff
@@ -42,9 +42,10 @@
 // Pack Limits
 #define MAX_PACK_CHG_CURR \
 	30 /* Pack-level charge limit: (MAX_CHG_CURR - 3.5A margin) × 3 cells in parallel */
-#define MAX_PACK_DISCHG_CURR \
-	(MAX_DISCHG_CURR *   \
-	 3) /* Pack-level discharge limit: MAX_DISCHG_CURR × 3 cells in parallel */
+// #define MAX_PACK_DISCHG_CURR \
+// 	(MAX_DISCHG_CURR *   \
+// 	 3) /* Pack-level discharge limit: MAX_DISCHG_CURR × 3 cells in parallel */
```
Comment (jr1221) [inline Core/Inc/bmsConfig.h]: undo

File: Core/Inc/bmsConfig.h [hunk 5]
```diff
@@ -53,25 +54,22 @@
...
-#define OCV_CURR_THRESH 0.0015 /* 1.5 mA */
+#define OCV_CURR_THRESH 0.5 /* in A */
 
 // Charging settings
-#define CHARGING_CURRENT    5
-#define CHARGE_SETL_TIMEOUT 60000 // 1 minute, may need adjustment
-#define CHARGE_SETL_TIMEUP  300000 // 5 minutes, may need adjustment
+#define CHARGING_CURRENT    4.0
+#define CHARGE_SETL_TIMEOUT 30000 // 1 minute, may need adjustment
+#define CHARGE_SETL_TIMEUP  120000 // 5 minutes, may need adjustment
```
Comment (jr1221) [inline Core/Inc/bmsConfig.h:61]: make issue for better charge algo

File: Core/Src/analyzer.c [hunk 6]
```diff
@@ -153,6 +226,9 @@ void calc_pack_temps(acc_data_t *bmsdata)
 				total_seg_temp /
 				((float)(NUM_CELLS_ALPHA + NUM_CELLS_BETA));
 			total_seg_temp = 0;
+			// if (c == 5) {
+			// 	printf("%f", se)
+			// }
```
Comment (jr1221) [inline Core/Src/analyzer.c]: remove

File: Core/Src/analyzer.c [hunk 7]
```diff
@@ -398,10 +490,38 @@ void calc_cont_ccl(acc_data_t *bmsdata)
 
 	bmsdata->cont_CCL = MAX_PACK_CHG_CURR * temp_cold_factor *
 			    temp_hot_factor * cell_volt_derate_factor;
+	bmsdata->cont_CCL = MAX_PACK_CHG_CURR;
```
Comment (jr1221) [inline Core/Src/analyzer.c]: undo

File: Core/Src/analyzer.c [hunk 8]
```diff
@@ -351,14 +442,15 @@ void calc_cont_dcl(acc_data_t *bmsdata)
 		scaled_dcl = MIN_DCL;
 	}
 
-	bmsdata->cont_DCL = scaled_dcl;
+	//bmsdata->cont_DCL = scaled_dcl;
+	bmsdata->cont_DCL = MAX_PACK_DISCHG_CURR;
```
Comment (jr1221) [inline Core/Src/analyzer.c]: undo

File: Core/Src/can_messages.c (hunk 1 above)
Comment (jr1221) [reply Core/Src/can_messages.c:260]: hmm it works idk why.  ig to be fixed

File: Core/Src/charging.c [hunk 9]
```diff
@@ -30,28 +30,27 @@ static float calc_cell_voltage_std(acc_data_t *data)
...
-	float thresh = avg - (STD_FACTOR * std);
 
 	/* Balance all cells above the threshold */
 	for (int chip = 0; chip < NUM_CHIPS; chip++) {
 		for (int cell = 0; cell < NUM_CELLS_ALPHA; cell++) {
 			/* Check if cell voltage is above (average - standard deviation) */
-			if (bmsdata->chip_data[chip].cell_voltages[cell] >
-			    thresh) {
+			if (bmsdata->chip_data[chip].open_cell_voltage[cell] >
+			    (low + thresh)) {
```
Comment (jr1221) [inline Core/Src/charging.c:53]: TODO: verify algorithm

File: Core/Src/compute.c (hunk 3 above)
Comment (jr1221) [reply Core/Src/compute.c:71]: its inverted compared to cerb

File: Core/Src/main.c [hunk 10]
```diff
@@ -339,9 +350,23 @@ int main(void)
   acc_data_t *acc_data = malloc(sizeof(acc_data_t));
   acc_data->is_charger_connected = false;
   acc_data->is_charging_enabled = false;
+  acc_data->pack_current = 0;
```
Comment (jr1221) [inline Core/Src/main.c:354]: note: if pack voltage parsing doesnt work then ocv=voltage

File: Core/Src/main.c [hunk 11]
```diff
@@ -339,9 +350,23 @@ int main(void)
...
+  acc_data->segment_average_temps[1] = 33.33;
+  acc_data->segment_average_temps[2] = 33.33;
+  acc_data->segment_average_temps[3] = 33.33;
+  acc_data->segment_average_temps[4] = 33.33;
+
+  acc_data->segment_average_volts[0] = 3.666;
+  acc_data->segment_average_volts[1] = 3.666;
+  acc_data->segment_average_volts[2] = 3.666;
+  acc_data->segment_average_volts[3] = 3.666;
+  acc_data->segment_average_volts[4] = 3.666;
```
Comment (jr1221) [inline Core/Src/main.c:369]: set defaults for averaging

File: Core/Src/shep_tasks.c [hunk 12]
```diff
@@ -39,21 +40,27 @@ void vGetSegmentData(void *pv_params)
...
-		if (current_state == CHARGING_STATE)
+		if (current_state == CHARGING_STATE) {
 			segment_retrieve_charging_data(bmsdata);
-		else
+		} else {
 			segment_retrieve_active_data(bmsdata);
-
-		if (DEBUG_MODE_ENABLED) {
-			segment_retrieve_debug_data(bmsdata);
+			if (DEBUG_MODE_ENABLED) {
```
Comment (jr1221) [inline Core/Src/shep_tasks.c:57]: this means in charging we always get debug data

File: Core/Src/shep_tasks.c [hunk 13]
```diff
@@ -171,7 +185,7 @@ void vDebugMode(void *pv_params)
 	acc_data_t *bmsdata = (acc_data_t *)pv_params;
 
 	// try to even everything out for a 1 Hz refresh rate
-	uint16_t time_per_chip = 1500 / NUM_CHIPS;
+	uint16_t time_per_chip = 750 / NUM_CHIPS;
```
Comment (jr1221) [inline Core/Src/shep_tasks.c]: remove?

File: Core/Src/shep_tasks.c [hunk 14]
```diff
@@ -211,7 +225,7 @@ void vDebugMode(void *pv_params)
 					 (cell + 1)) &
 						1);
 				// split half the time amongst the cells (over 2)
-				osDelay(time_per_chip / 2 / num_cells / 2);
+				osDelay(10);
```
Comment (jr1221) [inline Core/Src/shep_tasks.c:223]: TODO: improve frequencies

File: Core/Src/stateMachine.c [hunk 15]
```diff
@@ -127,6 +128,9 @@ void init_faulted(acc_data_t *bmsdata)
 {
 	// never balance when faulted
 	segment_disable_balancing(bmsdata);
+	// TODO fault fail
```
Comment (jr1221) [inline Core/Src/stateMachine.c]: remove comment

File: Core/Src/shep_tasks.c [hunk 16]
```diff
@@ -39,21 +40,27 @@ void vGetSegmentData(void *pv_params)
 	osDelay(500);
 
 	for (;;) {
+		// critical section so CAN does not block chip wake with ISR
+
+		segment_mute(bmsdata);
+
 		if (current_state == CHARGING_STATE) {
-			segment_mute(bmsdata);
+			osDelay(75);
```
Comment (caiodasilva2005) [inline Core/Src/shep_tasks.c:47]: Do we still need this osDelay? If so, we should just store the delay time of the thread in a variable and change it if we are in charging state. Right now we are osDelaying here and at the end of the thread

File: Core/Src/stateMachine.c [hunk 17]
```diff
@@ -364,6 +371,8 @@ bool sm_charging_check(acc_data_t *bmsdata)
 // check if balancing is allowed
 bool sm_balancing_check(acc_data_t *bmsdata)
 {
+	return false;
```
Comment (caiodasilva2005) [inline Core/Src/stateMachine.c:374]: remove

File: Core/Src/shep_tasks.c (hunk 16 above)
Comment (jr1221) [reply Core/Src/shep_tasks.c:47]: this osDelay is required to let the ADC settle after a segment_mute is called when balancing.

Comment (jr1221) [reply Core/Src/shep_tasks.c:47]: note to self: add a comment explaining that.

File: Core/Src/stateMachine.c (hunk 17 above)
Comment (jr1221) [reply Core/Src/stateMachine.c:374]: eh blocked by #239

File: Core/Src/can_messages.c (hunk 1 above)
Comment (jr1221) [reply Core/Src/can_messages.c:260]: bitstream_data is not a pointer but instead an array.  When referencing the array it is equal to &arr[0], but doing &arr implies this as well.  so its effectively the same.
